File: January 2026/07.py
from collections import defaultdict

class Solution:
    def countDistinct(self, arr, k):
        n = len(arr)
        if k > n:
            return []

        freq = defaultdict(int)
        result = []

        # Initialize the first window
        for i in range(k):
            freq[arr[i]] += 1
        result.append(len(freq))

        # Slide the window
        for i in range(k, n):
            # Remove the element going out of the window
            outgoing = arr[i - k]
            freq[outgoing] -= 1
            if freq[outgoing] == 0:
                del freq[outgoing]

            # Add the new element coming into the window
            incoming = arr[i]
            freq[incoming] += 1

            # Append the count of distinct elements
            result.append(len(freq))

        return result

File: January 2026/07.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_VAL 100001

void countDistinct(int arr[], int n, int k) {
    int freq[MAX_VAL] = {0};
    int distinct = 0;

    // First window
    for (int i = 0; i < k; ++i) {
        if (freq[arr[i]] == 0)
            distinct++;
        freq[arr[i]]++;
    }
    printf("%d ", distinct);

    // Slide window
    for (int i = k; i < n; ++i) {
        // Remove outgoing
        if (--freq[arr[i - k]] == 0)
            distinct--;

        // Add incoming
        if (freq[arr[i]] == 0)
            distinct++;
        freq[arr[i]]++;

        printf("%d ", distinct);
    }
}

int main() {
    int arr[] = {1, 2, 1, 3, 4, 2, 3};
    int k = 4;
    int n = sizeof(arr) / sizeof(arr[0]);
    countDistinct(arr, n, k);
    return 0;
}
